[23220] Request user password as confirmation for account changes

oliverguenther opened the pull request:
### Implementation approach

This feature will provide:
1.  An angular directive `<request-for-confirmation>`, that will wrap a form submit event and requests the confirmation for the given change.
2. An internal Rails filter to accept a challenge tag and user password. When password was confirmed, persists the tag along with a timestamp for the given user.

The form, rendered by rails, will output with a directive attribute when password confirmation is required. The password confirmation can be globally disabled. When we're unable to confirm the password for the current user (e.g., authenticated through OmniAuth), the dialog is skipped.

The directive decorates the form submission, displays the confirmation dialog (cf., screenshot) and extends the request with the password.

A before action will check for the confirmation with the given challenge and accept or decline the request.

**Design**

The directive will overlay a modal with the password confirmation form to be handled.

![step2](https://cloud.githubusercontent.com/assets/459462/17324612/affcc12c-58a7-11e6-921f-d4efefa5310b.png)
### Alternatives

Perform the confirmation with decorated actions in the Rails world. This would require redirecting / rewriting the actual form action and provide more potential code paths.
### Out of scope
- Providing the same functionality for users with externally provided authentication (through OmniAuth). 
- Restricting admin functionality (e.g., changing user details as a logged-in admin)
### ToDos
- [x] Consolidate the available modal implementations into a usable one. Extracted into https://github.com/opf/openproject/pull/4749
- [x] Implement the confirmation modal
- [X] Implement the confirmation filter
- [X] After discussing with @machisuji, replace the setting with a configuration variable
- [x] Decorate the user's profile actions with the confirmation
  - [X] For the core
  - [x] For Mobile-OTP (https://github.com/finnlabs/openproject-mobile_otp/pull/73)

https://community.openproject.com/work_packages/23220/activity

TeatroIO commented:
I've prepared a stage to preview changes. <a href='http://oliverguenther-feature-23220-password-confirmation.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com/' target='_blank'>Open stage</a> or <a href='http://oliverguenther-feature-23220-password-confirmation.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com//teatro.log' target='_blank'>view logs</a>.

oliverguenther committed:
Implement request-for-confirmation directive and modal.

oliverguenther committed:
Implement confirmation confern for my/account

oliverguenther committed:
Make global setting for password confirmation

Implements a check to request password confirmation only when:

1. `Setting.internal_password_confirmation` is set
2. The user is internally authenticated (`user.uses_external_authentication?` is false)

[ci skip]

oliverguenther committed:
Extend My Account specs

oliverguenther committed:
Replace setting with a configuration variable

oliverguenther committed:
Remove outdated legacy spec

machisuji commented:
👍
